New formula internal json v3 dependencies format

apainintheneck opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

See https://github.com/Homebrew/brew/issues/16410#issuecomment-1963079892

This implements a new way of storing dependencies in the internal formula JSON v3 format. The goal is that this format is easier to understand at a glance, easier to generate and easier to parse. I think with this PR two of those three goals have been achieved.

The parsing code ended up being a similar level of complexity to what we saw before. In the end, the JSON output will be slightly larger as mentioned in the thread linked above but the overall JSON size will still be much smaller than before. We're talking about half a megabyte difference when the total savings will end up being around 11 megabytes out of what was originally a 24 megabyte file (I haven't double-checked these numbers but they should be in the neighborhood and more details can be found in the thread).

The generation code is much simpler than before. Take a look at the old `Formula#dependencies_hash` method and the new `Formula#dependencies_list` method.

The biggest risk here is that it affects JSON generation for the public formula JSON v2 format or that the current JSON parsing logic doesn't work the same way it did before. If there are any problems with the internal JSON v3 parsing, no one's using it right now so it's not a big deal.

apainintheneck commented on Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json:
```diff
@@ -110,7 +91,29 @@
             "sha256": "ab49318d75eed3ee932c8e5add22f252ec0c852aad94945022877f926e93899f"
           }
         }
+      },
+      "dependencies": {
+        "cmake": {
+          "tags": [
+            "build"
+          ]
+        },
+        "python@3.12": {
+          "tags": [
+            "build"
+          ]
+        },
+        "llvm": {
+          "tags": [
+            "build",
+            "test"
+          ],
+          "uses_from_macos": null
+        },
+        "zlib": {
+          "uses_from_macos": null
+        }
```
It might make sense to add another formula that has uses from macOS bounds to make sure that's working correctly. Or I guess I could replace one of these examples too in a pinch.

MikeMcQuaid replied:
@apainintheneck Yeh, good idea 👍🏻

MikeMcQuaid commented:
> The biggest risk here is that it affects JSON generation for the public formula JSON v2 format or that the current JSON parsing logic doesn't work the same way it did before.

Might be nice to verify this by manually generating the formulae.brew.sh API output and diffing them?

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -2608,6 +2609,19 @@ def dependencies_hash
+    return if spec_symbol != :stable && spec_symbol != :head
+
+    send(spec_symbol)&.declared_deps&.each_with_object({}) do |dep, dep_hash|
+      next if dep.implicit? # Remove all implicit deps
```
Would be good to say why they are being removed rather than that they are

apainintheneck replied:
This comment was copied from the old logic. I'll do some digging and update it.

apainintheneck replied:
It seems like implicit dependencies are really only needed when building from source and are used when downloading and unpacking files. See `DependencyCollector` for more details.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -2608,6 +2609,19 @@ def dependencies_hash
+    send(spec_symbol)&.declared_deps&.each_with_object({}) do |dep, dep_hash|
+      next if dep.implicit? # Remove all implicit deps
+
+      dep_hash[dep.name] = {}.tap do |info|
```
Don't feel strongly but unclear if `tap` helps or harms readability here 🤔

apainintheneck replied:
Yeah, I tend to abuse `#tap` and `#then` a bit. I'll simplify it.

apainintheneck commented on Library/Homebrew/formula.rb:
```diff
@@ -2608,6 +2609,19 @@ def dependencies_hash
+  def dependencies_list(spec_symbol)
+    return if spec_symbol != :stable && spec_symbol != :head
+
+    send(spec_symbol)&.declared_deps&.each_with_object({}) do |dep, dep_hash|
```
I'll go with the following.

```rb
return unless (spec = public_send(spec_symbol))
```

MikeMcQuaid replied:
Works for me 👍🏻

apainintheneck commented:
I'm hoping that I've addressed all of the feedback in https://github.com/Homebrew/brew/pull/17153/commits/6ad02b8e340e163ce90beccb33d0f297d96e7b03. Let me know if there's anything I've missed.

It adds a new test for a formula with uses from macOS bounds, cleans up syntax and adds some more documentation.

apainintheneck commented:
> > The biggest risk here is that it affects JSON generation for the public formula JSON v2 format or that the current JSON parsing logic doesn't work the same way it did before.
> 
> Might be nice to verify this by manually generating the formulae.brew.sh API output and diffing them?

Yeah, I ran some extra tests locally and everything looks good to me. I'm not too worried about the generation side of things since that's relatively easy to test. The loading side is a bit more difficult but I didn't really touch anything that should affect the JSON v2 loading logic and those things are covered by some tests too. It was more just to let everyone know in case something goes haywire.

```console
$ brew api-readall-test --formula
Generating formulae API ... and mocking formula API loader.
Read core formula and saw no failures!
$ brew generate-api-diff --formula --stat
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
Switched to branch 'new-formula-json-dependencies-format'
 .../api/internal/v3/homebrew-core.json                                                                          | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
```

Note: It makes sense that the internal JSON v3 file is different after this change but everything else looks to be identical.

See my repo with dev commands for more info about each of them: https://github.com/apainintheneck/homebrew-dev-utils

MikeMcQuaid commented:
Thanks @apainintheneck!
